storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -0,0 +1,420 @@
+                    LOG.info("Failed to download basic resources for topology-id {}", _topologyId);
+                    Utils.forceDelete(tmproot);
+                    Utils.forceDelete(stormroot);
+                }
+            }
+            return null;
+        }
+    }
+    
+    private class DownloadBlobs implements Callable<Void> {
+        private final String _topologyId;
+
+        public DownloadBlobs(String topologyId) {
+            _topologyId = topologyId;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            String stormroot = ConfigUtils.supervisorStormDistRoot(_conf, _topologyId);
+            Map<String, Object> stormConf = ConfigUtils.readSupervisorStormConf(_conf, _topologyId);
+                
+            @SuppressWarnings("unchecked")
+            Map<String, Map<String, Object>> blobstoreMap = (Map<String, Map<String, Object>>) stormConf.get(Config.TOPOLOGY_BLOBSTORE_MAP);
+            String user = (String) stormConf.get(Config.TOPOLOGY_SUBMITTER_USER);
+            String topoName = (String) stormConf.get(Config.TOPOLOGY_NAME);
+
+            List<LocalResource> localResourceList = new ArrayList<>();
+            if (blobstoreMap != null) {
+                List<LocalResource> tmp = SupervisorUtils.blobstoreMapToLocalresources(blobstoreMap);
+                if (tmp != null) {
+                    localResourceList.addAll(tmp);
+                }
+            }
+
+            StormTopology stormCode = ConfigUtils.readSupervisorTopology(_conf, _topologyId, _fsOps);
+            List<String> dependencies = new ArrayList<>();
+            if (stormCode.is_set_dependency_jars()) {
+                dependencies.addAll(stormCode.get_dependency_jars());
+            }
+            if (stormCode.is_set_dependency_artifacts()) {
+                dependencies.addAll(stormCode.get_dependency_artifacts());
+            }
+            for (String dependency : dependencies) {
+                localResourceList.add(new LocalResource(dependency, false));
+            }
+            
+            if (!localResourceList.isEmpty()) {
+                File userDir = _localizer.getLocalUserFileCacheDir(user);
+                if (!userDir.exists()) {
+                    FileUtils.forceMkdir(userDir);
+                }
+                List<LocalizedResource> localizedResources = _localizer.getBlobs(localResourceList, user, topoName, userDir);
+                _fsOps.setupBlobPermissions(userDir, user);
+                for (LocalizedResource localizedResource : localizedResources) {
+                    File rsrcFilePath = new File(localizedResource.getFilePath());
+                    String keyName = rsrcFilePath.getName();
+                    String blobSymlinkTargetName = new File(localizedResource.getCurrentSymlinkPath()).getName();
+
+                    String symlinkName = null;
+                    Map<String, Object> blobInfo = blobstoreMap.get(keyName);
```

Comment:
If we use --jars or --dependencies and don't set blobstore map, blobstoreMap could be null.

Below code resolves NPE issue and make it works, but it might be better to clarify that it is better to include dependencies to blobstore map or keep this as it is.

```
if (blobstoreMap != null) {
                        Map<String, Object> blobInfo = blobstoreMap.get(keyName);
                        if (blobInfo != null && blobInfo.containsKey("localname")) {
                            symlinkName = (String) blobInfo.get("localname");
                        } else {
                            symlinkName = keyName;
                        }
                    } else {
                        // all things are from dependencies
                        symlinkName = keyName;
                    }
```